io.openems.shared.influxdb/src/io/openems/shared/influxdb/InfluxConnectorFluxQL.java

```diff
@@ -0,0 +1,414 @@
+			fromDate = fromDate.with(TemporalAdjusters.firstDayOfMonth());
+			if (!toDate.equals(toDate.with(TemporalAdjusters.firstDayOfMonth()))) {
+				toDate = toDate.with(TemporalAdjusters.lastDayOfMonth()).plusDays(1);
+			}
+		}
+
+		// handle empty call
+		if (channels.isEmpty()) {
+			return new TreeMap<>();
+		}
+
+		// prepare query
+		Flux flux = Flux.from(this.bucket) //
+				.range(fromDate.toInstant(), toDate.toInstant()) //
+				.filter(Restrictions.measurement().equal(InfluxConnector.MEASUREMENT));
+
+		if (influxEdgeId.isPresent()) {
+			flux = flux.filter(Restrictions.tag(OpenemsOEM.INFLUXDB_TAG).equal(influxEdgeId.get().toString()));
+		}
+
+		flux = flux.filter(InfluxConnectorFluxQL.toChannelAddressFieldList(channels)) //
+				.aggregateWindow(resolution.getValue(), resolution.getUnit(), "last") //
+				.difference(true);
+
+		var queryResult = this.executeQuery(flux);
+
+		return InfluxConnectorFluxQL.convertHistoricDataQueryResult(queryResult, fromDate, resolution);
+	}
+
+	/**
+	 * Queries historic data.
+	 *
+	 * @param influxEdgeId the unique, numeric Edge-ID; or Empty to query all Edges
+	 * @param fromDate     the From-Date
+	 * @param toDate       the To-Date
+	 * @param channels     the Channels to query
+	 * @param resolution   the resolution in seconds
+	 * @return the historic data as Map
+	 * @throws OpenemsException on error
+	 */
+	public SortedMap<ZonedDateTime, SortedMap<ChannelAddress, JsonElement>> queryHistoricData(
+			Optional<Integer> influxEdgeId, ZonedDateTime fromDate, ZonedDateTime toDate, Set<ChannelAddress> channels,
+			Resolution resolution) throws OpenemsNamedException {
+
+		// handle empty call
+		if (channels.isEmpty()) {
+			return new TreeMap<>();
+		}
+
+		// remove 5 minutes to prevent shifted timeline
+		var fromInstant = fromDate.toInstant().minus(5, ChronoUnit.MINUTES);
+
+		// prepare query
+		Flux flux = Flux.from(this.bucket) //
+				.range(fromInstant, toDate.toInstant()) //
+				.filter(Restrictions.measurement().equal(InfluxConnector.MEASUREMENT));
+
+		if (influxEdgeId.isPresent()) {
+			flux = flux.filter(Restrictions.tag(OpenemsOEM.INFLUXDB_TAG).equal(influxEdgeId.get().toString()));
+		}
+
+		flux = flux.filter(InfluxConnectorFluxQL.toChannelAddressFieldList(channels)) //
+				.aggregateWindow(resolution.getValue(), resolution.getUnit(), "mean");
+
+		// Execute query
+		var queryResult = this.executeQuery(flux);
+
+		return InfluxConnectorFluxQL.convertHistoricDataQueryResult(queryResult, fromDate, resolution);
+	}
+
+	/**
+	 * Queries the latest available channel values.
+	 *
+	 * @param influxEdgeId     the unique, numeric Edge-ID; or Empty to query all
+	 *                         Edges
+	 * @param channelAddresses the {@link ChannelAddress}es
+	 * @return a map of {@link ChannelAddress}es and values
+	 * @throws OpenemsException on error
+	 */
+	public Map<ChannelAddress, JsonElement> queryChannelValuesMap(Optional<Integer> influxEdgeId,
```

Comment:
i can only see 1 use and its within the class itself -> therefor should be private